Versioning policy (#1181)
ydb-platform/ydb-go-sdk: VERSIONING.md

# YDB-Go-SDK Versioning Policy

By adhering to these guidelines and exceptions, we aim to provide a stable and reliable development experience for our users while still allowing for innovation and improvement.

We endeavor to adhere to versioning guidelines as defined by [SemVer2.0.0](https://semver.org/).

We making the following exceptions to those guidelines:

> 1) **Experimental**

Author: ```suggestion
## Experimental
```

   - We use the `// Experimental` comment for new features in the `ydb-go-sdk`. 
   - Early adopters of newest feature can report bugs and imperfections in functionality. 
   - For fix this issues we can make broken changes in experimental API. 
   - We reserve the right to remove or modify these experimental features at any time, until the removal of the "Experimental" comment without increase of major part of version.

> 2) **Deprecated**

Author: ```suggestion
## Deprecated
```

   - We use the `// Deprecated` comment for deprecated features in the `ydb-go-sdk`.

>    - Usage of some entity marked with `// Deprecated` can be detected with linters such as [check-deprecated](https://github.com/black-06/check-deprecated). [staticcheck](https://github.com/dominikh/go-tools/tree/master/cmd/staticcheck) or [go-critic](https://github.com/go-critic/go-critic).

Author: ```suggestion
   - Usage of some entity marked with `// Deprecated` can be detected with linters such as [check-deprecated](https://github.com/black-06/check-deprecated), [staticcheck](https://github.com/dominikh/go-tools/tree/master/cmd/staticcheck) or [go-critic](https://github.com/go-critic/go-critic).
```

   - This helps to our users to soft decline to use the deprecated feature without any impact on their code.
   - Deprecated features will not be removed or changed for a minimum period of **six months** since the mark added.

>    - We reserve the right to remove or modify these deprecated features without increase to major part of version.

Author: ```suggestion
   - We reserve the right to remove or modify these deprecated features without increase of major part of version.
```

> 3) **Internals**

Author: ```suggestion
## Internals
```

1) **Experimental**
   - We use the `// Experimental` comment for new features in the `ydb-go-sdk`. 
   - Early adopters of newest feature can report bugs and imperfections in functionality. 
   - For fix this issues we can make broken changes in experimental API. 
   - We reserve the right to remove or modify these experimental features at any time, until the removal of the "Experimental" comment without increase of major part of version.
2) **Deprecated**
   - We use the `// Deprecated` comment for deprecated features in the `ydb-go-sdk`.
   - Usage of some entity marked with `// Deprecated` can be detected with linters such as [check-deprecated](https://github.com/black-06/check-deprecated). [staticcheck](https://github.com/dominikh/go-tools/tree/master/cmd/staticcheck) or [go-critic](https://github.com/go-critic/go-critic).
   - This helps to our users to soft decline to use the deprecated feature without any impact on their code.
   - Deprecated features will not be removed or changed for a minimum period of **six months** since the mark added.
   - We reserve the right to remove or modify these deprecated features without increase to major part of version.
3) **Internals**

>    - Some public packages of `ydb-go-sdk` relate to the internals.

Author: ```suggestion
   - Some public API of `ydb-go-sdk` relate to the internals.
```

   - We use the `// Internals` comment for public internals in the `ydb-go-sdk`.

>    - `ydb-go-sdk` internals can be changed at any time.

Author: ```suggestion
   - `ydb-go-sdk` internals can be changed at any time without increase of major part of version.
```

# YDB-Go-SDK Versioning Policy

By adhering to these guidelines and exceptions, we aim to provide a stable and reliable development experience for our users while still allowing for innovation and improvement.

We endeavor to adhere to versioning guidelines as defined by [SemVer2.0.0](https://semver.org/).

We making the following exceptions to those guidelines:
1) **Experimental**
   - We use the `// Experimental` comment for new features in the `ydb-go-sdk`. 
   - Early adopters of newest feature can report bugs and imperfections in functionality. 
   - For fix this issues we can make broken changes in experimental API. 

>    - We reserve the right to remove or modify these experimental features at any time, until the removal of the "Experimental" comment.

Author: ```suggestion
   - We reserve the right to remove or modify these experimental features at any time, until the removal of the "Experimental" comment without increase to major part of version.
```

2) **Deprecated**
   - We use the `// Deprecated` comment for deprecated features in the `ydb-go-sdk`.
   - Usage of some entity marked with `// Deprecated` can be detected with linters such as [check-deprecated](https://github.com/black-06/check-deprecated). [staticcheck](https://github.com/dominikh/go-tools/tree/master/cmd/staticcheck) or [go-critic](https://github.com/go-critic/go-critic).
   - This helps to our users to soft decline to use the deprecated feature without any impact on their code.

>    - Deprecated features will not be removed or changed for a minimum period of **six months**.

Reviewer: ```suggestion
   - Deprecated features will not be removed or changed for a minimum period of **six months** since the mark added.
```

>    - Deprecated features will not be removed or changed for a minimum period of **six months** since the mark added.

Author: ```suggestion
   - Deprecated features will not be removed or changed for a minimum period of **six months** since the mark added.
   - We reserve the right to remove or modify these deprecated features without increase to major part of version.
```

1) **Experimental**
   - We use the `// Experimental` comment for new features in the `ydb-go-sdk`. 
   - Early adopters of newest feature can report bugs and imperfections in functionality. 
   - For fix this issues we can make broken changes in experimental API. 
   - We reserve the right to remove or modify these experimental features at any time, until the removal of the "Experimental" comment without increase of major part of version.
2) **Deprecated**
   - We use the `// Deprecated` comment for deprecated features in the `ydb-go-sdk`.
   - Usage of some entity marked with `// Deprecated` can be detected with linters such as [check-deprecated](https://github.com/black-06/check-deprecated). [staticcheck](https://github.com/dominikh/go-tools/tree/master/cmd/staticcheck) or [go-critic](https://github.com/go-critic/go-critic).
   - This helps to our users to soft decline to use the deprecated feature without any impact on their code.
   - Deprecated features will not be removed or changed for a minimum period of **six months** since the mark added.
   - We reserve the right to remove or modify these deprecated features without increase to major part of version.

> 3) **Public internals**

Author: ```suggestion
3) **Unstable**
```

   - Some public packages of `ydb-go-sdk` relate to the internals.
   - `ydb-go-sdk` internals can be changed at any time.

>    - That's why versioning rules of [SemVer2.0.0](https://semver.org/) does not apply to the next packages:

Author: ```suggestion
   - That's why versioning rules of [SemVer2.0.0](https://semver.org/) does not apply to internals:
```

>      - [trace](https://github.com/ydb-platform/ydb-go-sdk/tree/master/trace)

Author: ```suggestion
```

     - [log](https://github.com/ydb-platform/ydb-go-sdk/tree/master/log)

>      - [metrics](https://github.com/ydb-platform/ydb-go-sdk/tree/master/metrics)

Author: ```suggestion
```

>    - List of stable supported metrics are stored in [metrics/README.md](https://github.com/ydb-platform/ydb-go-sdk/tree/master/metrics) and cover with tests

Author: ```suggestion
```
